crates/node/src/lib.rs

```diff
@@ -704,12 +707,111 @@ impl Node {
 
         Ok(outcome)
     }
-
     async fn handle_stream(
         &mut self,
-        _stream: calimero_network::stream::Stream,
+        mut stream: calimero_network::stream::Stream,
     ) -> eyre::Result<()> {
-        // TODO: implement catchup mechanism once storage is ready
+        let request = match stream.next().await {
+            Some(message) => match message {
+                Ok(message) => match serde_json::from_slice(&message.data)? {
+                    types::CatchupStreamMessage::Request(req) => req,
+                    message => {
+                        eyre::bail!("Unexpected message: {:?}", message)
+                    }
+                },
+                Err(err) => eyre::bail!(err),
+            },
+            None => {
+                eyre::bail!("Stream closed unexpectedly")
+            }
+        };
+
+        let handle = self.store.handle();
+
+        let meta_key = calimero_store::key::ContextMeta::new(request.context_id);
+        let Some(meta_value) = handle.get(&meta_key)? else {
+            let message = serde_json::to_vec(&types::CatchupError::ContextNotFound {
+                context_id: request.context_id,
+            })?;
+            stream
+                .send(calimero_network::stream::Message { data: message })
+                .await?;
+            return Ok(());
+        };
+
+        let mut hashes = VecDeque::new();
+        let mut key = calimero_store::key::ContextTransaction::new(
+            request.context_id,
+            meta_value.last_transaction_hash.into(),
+        );
+
+        let mut last_executed_transaction_hash_discovered = false;
+        while let Some(transaction) = handle.get(&key)? {
+            hashes.push_front(transaction.prior_hash);
+            if transaction.prior_hash == *request.last_executed_transaction_hash {
+                last_executed_transaction_hash_discovered = true;
+                break;
+            }
+
+            key = calimero_store::key::ContextTransaction::new(
+                request.context_id,
+                transaction.prior_hash.into(),
+            );
+        }
+
+        if !last_executed_transaction_hash_discovered {
+            let message = serde_json::to_vec(&types::CatchupError::TransactionNotFound {
```

Comment:
just noting that - depending on how the transactions are applied on receipt, we may need a different error for this case

`TransactionNotFound` can be quickly returned without traversing tx links

Just do a `get` on the last executed we're given so failure is instant and we don't spend a lot of time looking for it

This case, however, represents the potential for us to have transactions that aren't part of the canonical history

Which, should be managed such that it doesn't happen